QEI: Unable to (safely) define encoder pins as pull-up

I have this bit of code, which attempts to enable pull-up resistors on PB6 and PB7, and then construct a quadrature encoder interface on top of those pins using TIM4:

```rust
        let enc_a = gpiob.pb6.into_pull_up_input(&mut gpiob.crl);
        let enc_b = gpiob.pb7.into_pull_up_input(&mut gpiob.crl);
        let encoder = Timer::tim4(dp.TIM4, &clocks, &mut rcc.apb1).qei(
            (enc_a, enc_b),
            &mut afio.mapr,
            QeiOptions::default(),
        );
```

This fails because apparently the `qei()` method expects the pins to be floating instead:

```
error[E0277]: the trait bound `PB6<Input<PullUp>>: stm32f1xx_hal::gpio::Mode<Input<Floating>>` is not satisfied
  --> src\main.rs:95:68
   |
95 |         let encoder = Timer::tim4(dp.TIM4, &clocks, &mut rcc.apb1).qei(
   |                                                                    ^^^ the trait `stm32f1xx_hal::gpio::Mode<Input<Floating>>` is not implemented for `PB6<Input<PullUp>>`
   |
   = help: the following implementations were found:
             <PB6<MODE> as stm32f1xx_hal::gpio::Mode<MODE>>
   = note: required because of the requirements on the impl of `stm32f1xx_hal::pwm_input::Pins<Tim4NoRemap>` for `(PB6<Input<PullUp>>, PB7<Input<PullUp>>)`

error[E0277]: the trait bound `PB7<Input<PullUp>>: stm32f1xx_hal::gpio::Mode<Input<Floating>>` is not satisfied
  --> src\main.rs:95:68
   |
95 |         let encoder = Timer::tim4(dp.TIM4, &clocks, &mut rcc.apb1).qei(
   |                                                                    ^^^ the trait `stm32f1xx_hal::gpio::Mode<Input<Floating>>` is not implemented for `PB7<Input<PullUp>>`
   |
   = help: the following implementations were found:
             <PB7<MODE> as stm32f1xx_hal::gpio::Mode<MODE>>
   = note: required because of the requirements on the impl of `stm32f1xx_hal::pwm_input::Pins<Tim4NoRemap>` for `(PB6<Input<PullUp>>, PB7<Input<PullUp>>)`
```

Is there a way to do this safely and within the confines of the library? Note that this configuration seems to be valid as I have tested it using a C program generated in STM32CubeIDE and the behaviour is as I expect it.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - Added DMA receive support for `SPI`
 - Added `release` functions to SPI DMA
 - Add GPIOF/GPIOG support for high/xl density lines
+- Allow using `Input<PullUp>` and `Input<PullDown>` for timers' pwm and qei inputs.
 
 ### Fixed
 - Fix > 2 byte i2c reads

--- a/src/pwm_input.rs
+++ b/src/pwm_input.rs
@@ -12,7 +12,7 @@ use crate::pac::TIM4;
 use crate::pac::{TIM2, TIM3};
 
 use crate::afio::MAPR;
-use crate::gpio::{self, Floating, Input};
+use crate::gpio::{self, Input};
 use crate::rcc::{Clocks, GetBusFreq, RccBus};
 use crate::time::Hertz;
 use crate::timer::Timer;
@@ -21,11 +21,11 @@ pub trait Pins<REMAP> {}
 
 use crate::timer::sealed::{Ch1, Ch2, Remap};
 
-impl<TIM, REMAP, P1, P2> Pins<REMAP> for (P1, P2)
+impl<TIM, REMAP, P1, P2, MODE1, MODE2> Pins<REMAP> for (P1, P2)
 where
     REMAP: Remap<Periph = TIM>,
-    P1: Ch1<REMAP> + gpio::PinExt<Mode = Input<Floating>>,
-    P2: Ch2<REMAP> + gpio::PinExt<Mode = Input<Floating>>,
+    P1: Ch1<REMAP> + gpio::PinExt<Mode = Input<MODE1>>,
+    P2: Ch2<REMAP> + gpio::PinExt<Mode = Input<MODE2>>,
 {
 }
 

